utils/tlsmux/tlsmux.go

```diff
@@ -0,0 +1,73 @@
+package tlsmux
+
+import (
+	"crypto/tls"
+	"io"
+	"net"
+)
+
+type Conn struct {
+	net.Conn
+	initBuf []byte
+	initErr error
+}
+
+func (c *Conn) Read(b []byte) (int, error) {
+	if len(c.initBuf) == 0 {
+		return c.Conn.Read(b)
+	}
+	n := copy(b, c.initBuf)
+	if n == len(c.initBuf) {
+		c.initBuf = nil
+	} else {
+		c.initBuf = c.initBuf[n:]
+	}
+	if err := c.initErr; err != nil || n == len(b) {
+		c.initErr = nil
+		return n, err
+	}
+	nn, err := c.Conn.Read(b[n:])
+	return n + nn, err
+}
+
+type Listener struct {
+	net.Listener
+	TLSConfig *tls.Config
+}
+
+func NewListener(listener net.Listener, tlsConfig *tls.Config) *Listener {
+	return &Listener{
+		Listener:  listener,
+		TLSConfig: tlsConfig,
+	}
+}
+
+func (l *Listener) Accept() (net.Conn, error) {
+	c, err := l.Listener.Accept()
+	if err != nil {
+		return nil, err
+	}
+
+	// examine the first byte
+	buf := make([]byte, 1)
+	_, err = c.Read(buf)
+	if err != nil {
+		c.Close()
+		if err != io.EOF {
+			return nil, err
+		}
+	}
+
+	var conn net.Conn = &Conn{
+		Conn:    c,
+		initBuf: buf,
+		initErr: err,
+	}
```

Comment:
The way `Conn` is designed makes me think it's supposed to first send `buf` contents with the original `err`, then continue on to the rest of the stream. However, here we set `initBuf` to the whole `buf` we have, even though it may not have all its content set by the previous `Read`. I suggest setting `initBuf` to `buf[:n]` where `n` is (currently ignored) return value from `c.Read`.

Reply:
Oops, you're completely right, thanks for catching!